docs/developer/architecture/configurations.rst

```diff
@@ -73,3 +73,205 @@ As a MetalK8s expert, I can use ``kubectl`` command(s) in order to edit all
 settings that are exposed. The intent is to have a method / API that an expert
 could use, if the right CLI tool or GUI is not available or not functioning as
 expected.
+
+Design Choices
+--------------
+
+:term:`ConfigMap` is chosen as a unified data access and storage media for
+cluster and service configurations in a MetalK8s cluster based on the above
+requirements for the following reasons:
+
+* Ability to support Update operations on ConfigMap's with CLI and UI easily
+  using our already existing python kubernetes module.
+* Guarantee of adaptability and ease of changing the design and implementation
+  in cases were customer needs evolve rapidly.
+* ConfigMaps are stored in the :term:`etcd` database which is generally being
+  backed up. This ensures that user settings cannot be lost easily.
+
+How it works
+^^^^^^^^^^^^
+
+During the Bootstrap stages, when we are assertive that the K8s cluster is
+fully ready and available we could perform the following actions:
+
+  - Create and deploy ConfigMaps that hold cluster and service configurations
+    and pre-fill them with default values.
+  - Template service pods and deployments to consume configuration data
+    directly from the above deployed ConfigMaps
+
+This approach works because in a MetalK8s cluster, cluster and service
+configurations do not influence how the cluster behaves but rather the
+BootstrapConfiguration dictates the cluster behaviour. Also the creation of
+cluster and service configurations takes place before the respective consumers
+(MetalK8s addons) are deployed hence configuration data is always ready for
+consumers and not the other way round.
+
+
+**Using Salt states**
+
+Once a ConfigMap is updated by the user (say a user changes the number of
+replicas for Prometheus deployments to a new value), then perform the
+following actions:
+
+  - Apply a Salt states that reads the ConfigMap object, validates the schema
+    and checks the new values passed.
+
+Storage format
+~~~~~~~~~~~~~~
+
+A YAML(K8s-like) format was chosen to represent the data field instead of a
+flat key-value structure for the following reasons:
+
+ - YAML formatted configurations are easy to write and understand hence it will
+   be simpler for users to edit configurations.
+ - The YAML format benefits from bearing a schema version, which can be checked
+   before inferring the underlying service configuration if need arises.
+ - The YAML format utilizes white spaces to define variable hierarchies and as
+   such it is easier to manage with code mean while the flat key-value format
+   requires parsing of flat keys into hierarchical structures which can be
+   tedious.
```

Comment:
True, but not super clear. Maybe simply state that YAML **is** a format for describing hierarchical data structures, while using a flat structure would require a form of encoding (and then, decoding) of this hierarchical structure, which we know we'll need.